### PR TITLE
Add tests to clamping

### PR DESCRIPTION
**Pull request by rappie**

### CODE REVIEW

File: test/HelperClamp.t.sol [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+/**
+ * @dev Comprehensive tests for HelperClamp functionality including value clamping,
+ * bounds checking, logging, and edge cases with proper overflow handling.
+ * @author Perimeter <info@perimetersec.io>
+ */
+contract TestHelperClamp is Test, HelperClamp {
+    /**
+     * Tests for clamp(uint256, uint256, uint256)
+     */
+    function test_clamp_uint256_within_bounds() public {
```
Comment (ljz3) [inline test/HelperClamp.t.sol:18]: Don't need to do the casting

File: test/HelperClamp.t.sol [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+        // Test extreme value combinations
+        assertEq(this.clamp(maxUint, maxUint - 1, maxUint), maxUint);
+
+        // Cross-boundary tests - combinations of extreme values
+        assertEq(this.clamp(maxUint, uint256(0), maxUint - 1), uint256(0)); // Wraps to 0 due to modulo
+        assertEq(this.clamp(uint256(0), maxUint - 1, maxUint), maxUint - 1);
+    }
+
+    function testFuzz_clamp_uint256(uint256 value, uint256 low, uint256 high) public {
+        vm.assume(low <= high);
```
Comment (ljz3) [inline test/HelperClamp.t.sol:106]: We should have this case revert with a custom error

File: test/HelperClamp.t.sol [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+    function test_clamp_int256_with_logging() public {
+        vm.expectEmit(true, true, true, true);
+        emit Clamped("Clamping value 15 to -6");
+
+        int256 result = this.clamp(int256(15), int256(-10), int256(10));
+        assertEq(result, int256(-6));
+    }
+
+    function test_clamp_int256_invalid_range() public {
```
Comment (ljz3) [inline test/HelperClamp.t.sol:171]: This should be throwing an error

File: test/HelperClamp.t.sol [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+     * Tests for clampGt(int256, int256) - calls clamp(a, b + 1, type(int256).max)
+     */
+    function test_clampGt_int256_basic_behavior() public {
+        // Test that clampGt(a, b) equivalent to clamp(a, b + 1, type(int256).max)
+        int256 b = int256(10);
+        assertEq(this.clampGt(int256(5), b), this.clamp(int256(5), b + 1, type(int256).max));
+        assertEq(this.clampGt(int256(15), b), this.clamp(int256(15), b + 1, type(int256).max));
+    }
+
+    function test_clampGt_int256_overflow_cases() public {
```
Comment (ljz3) [inline test/HelperClamp.t.sol:383]: This should revert with a custom error

File: test/HelperClamp.t.sol [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+        // Test overflow case: clampGt(value, type(int256).max) calls clamp(value, type(int256).max + 1, type(int256).max)
+        // This should overflow when computing b + 1
+        vm.expectRevert();
+        this.clampGt(int256(100), type(int256).max);
+    }
+
+    // Note: int256 fuzz test omitted due to inherent overflow in extreme bounds
```
Comment (ljz3) [inline test/HelperClamp.t.sol:390]: This should not be omitted

File: test/HelperClamp.t.sol [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+    /**
+     * Tests for clampGt(uint256, uint256) - calls clamp(a, b + 1, type(uint256).max)
+     */
+    function test_clampGt_uint256_basic_behavior() public {
+        // Test that clampGt(a, b) equivalent to clamp(a, b + 1, type(uint256).max)
+        assertEq(this.clampGt(uint256(5), uint256(10)), this.clamp(uint256(5), uint256(11), type(uint256).max));
+        assertEq(this.clampGt(uint256(15), uint256(10)), this.clamp(uint256(15), uint256(11), type(uint256).max));
+    }
+
+    function test_clampGt_uint256_overflow_cases() public {
```
Comment (ljz3) [inline test/HelperClamp.t.sol:352]: This should revert with a custom error

File: test/HelperClamp.t.sol [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+        // Result should always be <= b
+        assertTrue(result <= b);
+
+        // If a was already <= b, result should equal a
+        if (a <= b) {
+            assertEq(result, a);
+        }
+    }
+
+    // Note: int256 fuzz test omitted due to inherent overflow in extreme bounds
```
Comment (ljz3) [inline test/HelperClamp.t.sol:323]: This should not be omitted

File: test/HelperClamp.t.sol [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+    }
+
+    function test_clampLt_int256_overflow_cases() public {
+        // Test overflow case: clampLt(value, type(int256).min) calls clamp(value, type(int256).min, type(int256).min - 1)
+        // This should overflow when computing b - 1
+        vm.expectRevert();
+        this.clampLt(int256(100), type(int256).min);
+    }
+
+    // Note: int256 fuzz test omitted due to inherent overflow in extreme bounds
```
Comment (ljz3) [inline test/HelperClamp.t.sol:298]: This should not be omitted

File: test/HelperClamp.t.sol [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+
+        uint256 result = this.clamp(uint256(15), uint256(0), uint256(9));
+        assertEq(result, uint256(5));
+    }
+
+    function test_clamp_uint256_invalid_range() public {
+        // When low > high, the range calculation will underflow/wrap around
+        // For uint256: high - low + 1 will wrap to a very large number
+        // This tests the behavior with invalid ranges - but we expect it to revert
+        vm.expectRevert(); // The function should revert due to underflow in range calculation
```
Comment (ljz3) [inline test/HelperClamp.t.sol:73]: Should revert with a custom error, not over/underflow

**Review by ljz3 — CHANGES_REQUESTED** [review]

File: test/HelperClamp.t.sol (hunk 1 above)
Comment (rappie) [reply test/HelperClamp.t.sol:18]: i thought the same but it seems to be necessary

```
Error (6675): Member "clamp" not unique after argument-dependent lookup in contract TestHelperClamp.
  --> test/HelperClamp.t.sol:26:18:
   |
26 |         assertEq(this.clamp(5, 0, 10), 5);
   |                  ^^^^^^^^^^
```

File: src/helpers/HelperClamp.sol [hunk 10]
```diff
@@ -103,144 +109,194 @@ abstract contract HelperClamp is HelperAssert {
...
+            return value;
+        }
+
+        // At this point: value is outside [low, high] and needs wrapping
+        uint256 ans;
+
+        if (low == high) {
+            // Edge case: Single-value range - everything maps to the same value
+            // Avoids potential division by zero in modulo calculation
+            ans = low;
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:148]: This is not necessary, it is impossible for the denominator in the modulo to be 0.

File: src/helpers/HelperClamp.sol [hunk 11]
```diff
@@ -103,144 +109,194 @@ abstract contract HelperClamp is HelperAssert {
...
      */
-    function clamp(int256 value, int256 low, int256 high, bool enableLogs) public returns (int256) {
-        if (value < low || value > high) {
-            int256 range = high - low + 1;
-            int256 clamped = (value - low) % (range);
-            if (clamped < 0) clamped += range;
-            int256 ans = low + clamped;
-            if (enableLogs) logClamp(value, ans);
-            return ans;
+    function clamp(int128 value, int128 low, int128 high, bool enableLogs) public returns (int128) {
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:193]: I would rather have the input be int256 and then safe cast it to int256, just so that it's less annoying to use

File: src/helpers/HelperClamp.sol [hunk 12]
```diff
@@ -103,144 +109,194 @@ abstract contract HelperClamp is HelperAssert {
...
+        }
+
+        // At this point: value is outside [low, high] and needs wrapping
+        int128 ans;
+
+        if (low == high) {
+            // Single-value range: everything maps to the same value
+            // Avoids potential division by zero in modulo calculation
+            ans = low;
+        } else {
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:209]: Not necessary, as this case is not possible.

File: src/helpers/HelperClamp.sol [hunk 13]
```diff
@@ -103,144 +109,194 @@ abstract contract HelperClamp is HelperAssert {
...
      */
-    function clampGt(int256 a, int256 b, bool enableLogs) public returns (int256) {
-        if (!(a > b)) {
-            int256 value = b + 1;
-            if (enableLogs) logClamp(a, value);
-            return value;
-        } else {
-            return a;
-        }
+    function clampGt(int128 a, int128 b, bool enableLogs) public returns (int128) {
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:283]: I would rather have the input be int256 and then safe cast it to int256, just so that it's less annoying to use

File: src/helpers/HelperClamp.sol [hunk 14]
```diff
@@ -103,144 +109,194 @@ abstract contract HelperClamp is HelperAssert {
...
      * @dev Clamps signed integer to be greater than or equal to specified value with optional logging.
      */
-    function clampGte(int256 a, int256 b, bool enableLogs) public returns (int256) {
-        if (!(a > b)) {
-            int256 value = b;
-            if (enableLogs) logClamp(a, value);
-            return value;
-        }
-        return a;
+    function clampGte(int128 a, int128 b, bool enableLogs) public returns (int128) {
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:298]: I would rather have the input be int256 and then safe cast it to int256, just so that it's less annoying to use

File: test/HelperClamp.t.sol (hunk 1 above)
Comment (ljz3) [reply test/HelperClamp.t.sol:17]: Ah yes, because it could be the int version as well

File: src/helpers/HelperClamp.sol (hunk 11 above)
Comment (rappie) [reply src/helpers/HelperClamp.sol:193]: do you mean safe casting it to int128 and reverting if this fails?

File: src/helpers/HelperClamp.sol [hunk 15]
```diff
@@ -36,21 +41,21 @@ abstract contract HelperClamp is HelperAssert {
      * @param high The maximum bound (inclusive)
      * @return The clamped value
      */
-    function clamp(int256 value, int256 low, int256 high) public returns (int256) {
+    function clamp(int256 value, int256 low, int256 high) public returns (int128) {
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:44]: Should still return int256, can just re-upcast it

File: src/helpers/HelperClamp.sol [hunk 16]
```diff
@@ -36,21 +41,21 @@ abstract contract HelperClamp is HelperAssert {
...
     function clampLt(uint256 a, uint256 b) public returns (uint256) {
-        return clampLt(a, b);
+        return clampLt(a, b, true);
     }
 
     /**
      * @dev Clamps signed integer to be less than specified value.
      */
-    function clampLt(int256 a, int256 b) public returns (int256) {
+    function clampLt(int256 a, int256 b) public returns (int128) {
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:58]: Should still return int256

File: src/helpers/HelperClamp.sol [hunk 17]
```diff
@@ -64,7 +69,7 @@ abstract contract HelperClamp is HelperAssert {
     /**
      * @dev Clamps signed integer to be less than or equal to specified value.
      */
-    function clampLte(int256 a, int256 b) public returns (int256) {
+    function clampLte(int256 a, int256 b) public returns (int128) {
```
Comment (ljz3) [inline src/helpers/HelperClamp.sol:72]: Should still return int256

**Review by ljz3 — CHANGES_REQUESTED** [review]
